Use CSS tokenizer for parsing standalone pseudo-elements

annevk opened the pull request:
#### 6fc1d6cf00094896d8b7a58d7dcd0aa1efdee97c
<pre>
Use CSS tokenizer for parsing standalone pseudo-elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=266979">https://bugs.webkit.org/show_bug.cgi?id=266979</a>

Reviewed by Tim Nguyen.

This ensures we do the correct thing for CSS escapes and also paves the
path for supporting FunctionToken and potentially deduplicating with
CSSSelectorParser::consumePseudo() in the future.

272607@main added additional test coverage for various token inputs.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parseStandalonePseudoElement):
* Source/WebCore/css/CSSSelector.h:

Canonical link: <a href="https://commits.webkit.org/272649@main">https://commits.webkit.org/272649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc2edbc598f6a331c3abb15368190871520d8d4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32496 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11234 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35051 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29390 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13587 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8433 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28916 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8245 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8389 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28970 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36387 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29534 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29395 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34518 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8521 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6474 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32372 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10188 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7570 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9140 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9101 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -358,22 +359,32 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
+        return std::nullopt;
+    auto pseudoElement = parsePseudoElement(token.value(), context);
+    if (!pseudoElement)
+        return std::nullopt;
```
Should you check the end of the range here? I suppose you want to properly reject `::foo bar`

nt1m replied:
(This should probably be tested fwiw)

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -334,22 +335,33 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
+        auto pseudoClassOrElement = parsePseudoClassAndCompatibilityElementString(token.value());
         if (!pseudoClassOrElement.compatibilityPseudoElement)
             return std::nullopt;
         return pseudoId(*pseudoClassOrElement.compatibilityPseudoElement);
```
We should ASSERT() here that the compatibility pseudo element is enabled btw. Like in `MutableCSSSelector::parsePseudoClassSelector`

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -334,22 +335,33 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
+    auto pseudoElement = parsePseudoElement(token.value(), context);
+    if (!pseudoElement)
+        return std::nullopt;
+    return pseudoId(*pseudoElement);
```
```suggestion
    if (auto pseudoElement = parsePseudoElement(token.value(), context))
        return pseudoId(*pseudoElement);
    return std::nullopt;
```

I don't have a strong preference, but you can write it like this, which is slightly shorter

annevk replied:
Thanks, big fan of that construct.

annevk committed:
Use CSS tokenizer for parsing standalone pseudo-elements
https://bugs.webkit.org/show_bug.cgi?id=266979

Reviewed by Tim Nguyen.

This ensures we do the correct thing for CSS escapes and also paves the
path for supporting FunctionToken and potentially deduplicating with
CSSSelectorParser::consumePseudo() in the future.

272607@main added additional test coverage for various token inputs.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parseStandalonePseudoElement):
* Source/WebCore/css/CSSSelector.h:

Canonical link: https://commits.webkit.org/272649@main
